// @apple/app-store-server-library/dist/models/ConsumptionRequestReason.js
"use strict";
// Copyright (c) 2024 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsumptionRequestReasonValidator = exports.ConsumptionRequestReason = void 0;
const Validator_1 = require("./Validator");
/**
 * The customer-provided reason for a refund request.
 *
 * {@link https://developer.apple.com/documentation/appstoreservernotifications/consumptionrequestreason consumptionRequestReason}
 */
var ConsumptionRequestReason;
(function (ConsumptionRequestReason) {
    ConsumptionRequestReason["UNINTENDED_PURCHASE"] = "UNINTENDED_PURCHASE";
    ConsumptionRequestReason["FULFILLMENT_ISSUE"] = "FULFILLMENT_ISSUE";
    ConsumptionRequestReason["UNSATISFIED_WITH_PURCHASE"] = "UNSATISFIED_WITH_PURCHASE";
    ConsumptionRequestReason["LEGAL"] = "LEGAL";
    ConsumptionRequestReason["OTHER"] = "OTHER";
})(ConsumptionRequestReason || (exports.ConsumptionRequestReason = ConsumptionRequestReason = {}));
class ConsumptionRequestReasonValidator extends Validator_1.StringValidator {
}
exports.ConsumptionRequestReasonValidator = ConsumptionRequestReasonValidator;
