// @apple/app-store-server-library/dist/models/ConsumptionStatus.js
"use strict";
// Copyright (c) 2023 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsumptionStatusValidator = exports.ConsumptionStatus = void 0;
const Validator_1 = require("./Validator");
/**
 * A value that indicates the extent to which the customer consumed the in-app purchase.
 *
 * {@link https://developer.apple.com/documentation/appstoreserverapi/consumptionstatus consumptionStatus}
 */
var ConsumptionStatus;
(function (ConsumptionStatus) {
    ConsumptionStatus[ConsumptionStatus["UNDECLARED"] = 0] = "UNDECLARED";
    ConsumptionStatus[ConsumptionStatus["NOT_CONSUMED"] = 1] = "NOT_CONSUMED";
    ConsumptionStatus[ConsumptionStatus["PARTIALLY_CONSUMED"] = 2] = "PARTIALLY_CONSUMED";
    ConsumptionStatus[ConsumptionStatus["FULLY_CONSUMED"] = 3] = "FULLY_CONSUMED";
})(ConsumptionStatus || (exports.ConsumptionStatus = ConsumptionStatus = {}));
class ConsumptionStatusValidator extends Validator_1.NumberValidator {
}
exports.ConsumptionStatusValidator = ConsumptionStatusValidator;
