// @apple/app-store-server-library/dist/models/DeliveryStatus.js
"use strict";
// Copyright (c) 2025 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeliveryStatusValidator = exports.DeliveryStatus = void 0;
const Validator_1 = require("./Validator");
/**
 * A value that indicates whether the app successfully delivered an In-App Purchase that works properly.
 *
 * {@link https://developer.apple.com/documentation/appstoreserverapi/deliverystatus deliveryStatus}
 */
var DeliveryStatus;
(function (DeliveryStatus) {
    DeliveryStatus["DELIVERED"] = "DELIVERED";
    DeliveryStatus["UNDELIVERED_QUALITY_ISSUE"] = "UNDELIVERED_QUALITY_ISSUE";
    DeliveryStatus["UNDELIVERED_WRONG_ITEM"] = "UNDELIVERED_WRONG_ITEM";
    DeliveryStatus["UNDELIVERED_SERVER_OUTAGE"] = "UNDELIVERED_SERVER_OUTAGE";
    DeliveryStatus["UNDELIVERED_OTHER"] = "UNDELIVERED_OTHER";
})(DeliveryStatus || (exports.DeliveryStatus = DeliveryStatus = {}));
class DeliveryStatusValidator extends Validator_1.StringValidator {
}
exports.DeliveryStatusValidator = DeliveryStatusValidator;
