// @apple/app-store-server-library/dist/models/DeliveryStatusV1.js
"use strict";
// Copyright (c) 2023 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeliveryStatusV1Validator = exports.DeliveryStatusV1 = void 0;
const Validator_1 = require("./Validator");
/**
 * A value that indicates whether the app successfully delivered an in-app purchase that works properly.
 *
 * {@link https://developer.apple.com/documentation/appstoreserverapi/deliverystatusv1 deliveryStatus}
 * @deprecated Use {@link DeliveryStatus} instead.
 */
var DeliveryStatusV1;
(function (DeliveryStatusV1) {
    DeliveryStatusV1[DeliveryStatusV1["DELIVERED_AND_WORKING_PROPERLY"] = 0] = "DELIVERED_AND_WORKING_PROPERLY";
    DeliveryStatusV1[DeliveryStatusV1["DID_NOT_DELIVER_DUE_TO_QUALITY_ISSUE"] = 1] = "DID_NOT_DELIVER_DUE_TO_QUALITY_ISSUE";
    DeliveryStatusV1[DeliveryStatusV1["DELIVERED_WRONG_ITEM"] = 2] = "DELIVERED_WRONG_ITEM";
    DeliveryStatusV1[DeliveryStatusV1["DID_NOT_DELIVER_DUE_TO_SERVER_OUTAGE"] = 3] = "DID_NOT_DELIVER_DUE_TO_SERVER_OUTAGE";
    DeliveryStatusV1[DeliveryStatusV1["DID_NOT_DELIVER_DUE_TO_IN_GAME_CURRENCY_CHANGE"] = 4] = "DID_NOT_DELIVER_DUE_TO_IN_GAME_CURRENCY_CHANGE";
    DeliveryStatusV1[DeliveryStatusV1["DID_NOT_DELIVER_FOR_OTHER_REASON"] = 5] = "DID_NOT_DELIVER_FOR_OTHER_REASON";
})(DeliveryStatusV1 || (exports.DeliveryStatusV1 = DeliveryStatusV1 = {}));
class DeliveryStatusV1Validator extends Validator_1.NumberValidator {
}
exports.DeliveryStatusV1Validator = DeliveryStatusV1Validator;
