// @apple/app-store-server-library/dist/models/ExpirationIntent.js
"use strict";
// Copyright (c) 2023 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpirationIntentValidator = exports.ExpirationIntent = void 0;
const Validator_1 = require("./Validator");
/**
 * The reason an auto-renewable subscription expired.
 *
 * {@link https://developer.apple.com/documentation/appstoreserverapi/expirationintent expirationIntent}
 */
var ExpirationIntent;
(function (ExpirationIntent) {
    ExpirationIntent[ExpirationIntent["CUSTOMER_CANCELLED"] = 1] = "CUSTOMER_CANCELLED";
    ExpirationIntent[ExpirationIntent["BILLING_ERROR"] = 2] = "BILLING_ERROR";
    ExpirationIntent[ExpirationIntent["CUSTOMER_DID_NOT_CONSENT_TO_PRICE_INCREASE"] = 3] = "CUSTOMER_DID_NOT_CONSENT_TO_PRICE_INCREASE";
    ExpirationIntent[ExpirationIntent["PRODUCT_NOT_AVAILABLE"] = 4] = "PRODUCT_NOT_AVAILABLE";
    ExpirationIntent[ExpirationIntent["OTHER"] = 5] = "OTHER";
})(ExpirationIntent || (exports.ExpirationIntent = ExpirationIntent = {}));
class ExpirationIntentValidator extends Validator_1.NumberValidator {
}
exports.ExpirationIntentValidator = ExpirationIntentValidator;
