// @apple/app-store-server-library/dist/models/ExtendReasonCode.js
"use strict";
// Copyright (c) 2023 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtendReasonCodeValidator = exports.ExtendReasonCode = void 0;
const Validator_1 = require("./Validator");
/**
 * The code that represents the reason for the subscription-renewal-date extension.
 *
 * {@link https://developer.apple.com/documentation/appstoreserverapi/extendreasoncode extendReasonCode}
 */
var ExtendReasonCode;
(function (ExtendReasonCode) {
    ExtendReasonCode[ExtendReasonCode["UNDECLARED"] = 0] = "UNDECLARED";
    ExtendReasonCode[ExtendReasonCode["CUSTOMER_SATISFACTION"] = 1] = "CUSTOMER_SATISFACTION";
    ExtendReasonCode[ExtendReasonCode["OTHER"] = 2] = "OTHER";
    ExtendReasonCode[ExtendReasonCode["SERVICE_ISSUE_OR_OUTAGE"] = 3] = "SERVICE_ISSUE_OR_OUTAGE";
})(ExtendReasonCode || (exports.ExtendReasonCode = ExtendReasonCode = {}));
class ExtendReasonCodeValidator extends Validator_1.NumberValidator {
}
exports.ExtendReasonCodeValidator = ExtendReasonCodeValidator;
