// @apple/app-store-server-library/dist/models/LifetimeDollarsPurchased.js
"use strict";
// Copyright (c) 2023 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LifetimeDollarsPurchasedValidator = exports.LifetimeDollarsPurchased = void 0;
const Validator_1 = require("./Validator");
/**
 * A value that indicates the total amount, in USD, of in-app purchases the customer has made in your app, across all platforms.
 *
 * {@link https://developer.apple.com/documentation/appstoreserverapi/lifetimedollarspurchased lifetimeDollarsPurchased}
 */
var LifetimeDollarsPurchased;
(function (LifetimeDollarsPurchased) {
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["UNDECLARED"] = 0] = "UNDECLARED";
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["ZERO_DOLLARS"] = 1] = "ZERO_DOLLARS";
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["ONE_CENT_TO_FORTY_NINE_DOLLARS_AND_NINETY_NINE_CENTS"] = 2] = "ONE_CENT_TO_FORTY_NINE_DOLLARS_AND_NINETY_NINE_CENTS";
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["FIFTY_DOLLARS_TO_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS"] = 3] = "FIFTY_DOLLARS_TO_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS";
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["ONE_HUNDRED_DOLLARS_TO_FOUR_HUNDRED_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS"] = 4] = "ONE_HUNDRED_DOLLARS_TO_FOUR_HUNDRED_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS";
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["FIVE_HUNDRED_DOLLARS_TO_NINE_HUNDRED_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS"] = 5] = "FIVE_HUNDRED_DOLLARS_TO_NINE_HUNDRED_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS";
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["ONE_THOUSAND_DOLLARS_TO_ONE_THOUSAND_NINE_HUNDRED_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS"] = 6] = "ONE_THOUSAND_DOLLARS_TO_ONE_THOUSAND_NINE_HUNDRED_NINETY_NINE_DOLLARS_AND_NINETY_NINE_CENTS";
    LifetimeDollarsPurchased[LifetimeDollarsPurchased["TWO_THOUSAND_DOLLARS_OR_GREATER"] = 7] = "TWO_THOUSAND_DOLLARS_OR_GREATER";
})(LifetimeDollarsPurchased || (exports.LifetimeDollarsPurchased = LifetimeDollarsPurchased = {}));
class LifetimeDollarsPurchasedValidator extends Validator_1.NumberValidator {
}
exports.LifetimeDollarsPurchasedValidator = LifetimeDollarsPurchasedValidator;
